### PR TITLE
feat: add Catalog and Schema Manager

### PR DESCRIPTION
**Pull request by WenyXu**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

1. Add Range Pagination Stream
2. Add Catalog and Schema Manager

## Checklist

- [x]  I have written the necessary rustdoc comments.
- [x]  I have added the necessary unit tests and integration tests.

## Refer to a related PR or issue link (optional)
- #2024 
- #1846

### CODE REVIEW

File: src/common/meta/src/util.rs [hunk 1]
```diff
@@ -27,10 +27,27 @@ pub fn get_prefix_end_key(key: &[u8]) -> Vec<u8> {
     vec![0]
 }
 
+/// Get next prefix key of `key`.
+#[inline]
+pub fn get_next_prefix_key(key: &[u8]) -> Vec<u8> {
+    let mut key = Vec::from(key);
```
Comment (MichaelScofield) [inline src/common/meta/src/util.rs:33]: Maybe more efficient:
```suggestion
    let mut key = Vec::with_capacity(key.len() + 1);
    key.extend_from_slice(key);
```

File: src/common/meta/src/range_stream.rs [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+    Error,
+}
+
+pub const DEFAULT_PAGE_SIZE: usize = 512;
+
+struct PaginationStreamFactory {
+    kv: KvBackendRef,
+    /// key is the first key for the range, If range_end is not given, the
+    /// request only looks up key.
+    pub key: Vec<u8>,
```
Comment (MichaelScofield) [inline src/common/meta/src/range_stream.rs:48]: suggest just store the entire `RangeRequest` to make codes concise

File: src/common/meta/src/key/catalog_name.rs [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+        resp.handle(|_| Ok(()))
+    }
+
+    fn build_create_req(schema: CatalogNameKey<'_>) -> Result<CompareAndPutRequest> {
+        let raw_key = schema.as_raw_key();
+
+        Ok(CompareAndPutRequest::new()
+            .with_key(raw_key)
+            .with_expect(vec![])
+            .with_expect(CatalogNameValue.try_as_raw_value()?))
```
Comment (MichaelScofield) [inline src/common/meta/src/key/catalog_name.rs:146]:
```suggestion
```

File: src/common/meta/src/key/schema_name.rs [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+    pub fn new(catalog: &'a str, schema: &'a str) -> Self {
+        Self { catalog, schema }
+    }
+
+    pub fn range_start_key(catalog: &str) -> String {
+        format!("{}/{}/", SCHEMA_NAME_KEY_PREFIX, catalog)
+    }
+}
+
+impl<'a> From<&'a TableName> for SchemaNameKey<'a> {
```
Comment (MichaelScofield) [inline src/common/meta/src/key/schema_name.rs:50]: same as above

File: src/common/meta/src/key/schema_name.rs [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+        resp.handle(|_| Ok(()))
+    }
+
+    fn build_create_req(schema: SchemaNameKey<'_>) -> Result<CompareAndPutRequest> {
+        let raw_key = schema.as_raw_key();
+
+        Ok(CompareAndPutRequest::new()
+            .with_key(raw_key)
+            .with_expect(vec![])
+            .with_expect(SchemaNameValue.try_as_raw_value()?))
```
Comment (MichaelScofield) [inline src/common/meta/src/key/schema_name.rs:144]: why two expects?

Comment (WenyXu) [reply src/common/meta/src/key/schema_name.rs:144]: typo 🥲

File: src/common/meta/src/range_stream.rs (hunk 2 above)
Comment (WenyXu) [reply src/common/meta/src/range_stream.rs:48]: However, The `limit` in the `RangeRequest` isn't the `page_size` in the `PaginationStreamFactory`.

**Comment by WenyXu** [comment]
@fengjiachun @MichaelScofield PTAL

File: src/common/meta/src/key/catalog_name.rs [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+    }
+}
+
+impl<'a, 'b: 'a> From<&'b SchemaNameKey<'b>> for CatalogNameKey<'a> {
+    fn from(value: &'b SchemaNameKey) -> Self {
+        Self {
+            catalog: value.catalog,
+        }
+    }
+}
```
[the remark this answers is not in the document]
Comment (MichaelScofield) [reply src/common/meta/src/key/catalog_name.rs]: Still has `impl<'a> From<&'a TableName> for CatalogNameKey<'a> {`

File: src/common/meta/src/key/catalog_name.rs [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+impl CatalogManager {
+    pub fn new(kv_backend: KvBackendRef) -> Self {
+        Self { kv_backend }
+    }
+
+    /// Creates `CatalogNameKey`, throws an error if key exists.
+    pub async fn create(&self, catalog: CatalogNameKey<'_>) -> Result<()> {
+        let resp = self
+            .kv_backend
+            .compare_and_put(Self::build_create_req(catalog)?)
```
Comment (MichaelScofield) [inline src/common/meta/src/key/catalog_name.rs:107]: On second thought, do we really need to use cas here? Catalog cannot be renamed. I think we can simply use "put" here, and do not need `create_if_not_exists`.

File: src/common/meta/src/key/schema_name.rs [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+impl SchemaManager {
+    pub fn new(kv_backend: KvBackendRef) -> Self {
+        Self { kv_backend }
+    }
+
+    /// Creates `SchemaNameKey`, throws an error if key exists.
+    pub async fn create(&self, schema: SchemaNameKey<'_>) -> Result<()> {
+        let resp = self
+            .kv_backend
+            .compare_and_put(Self::build_create_req(schema)?)
```
Comment (MichaelScofield) [inline src/common/meta/src/key/schema_name.rs:104]: Same as above.

**Comment by WenyXu** [comment]
> Is this a breaking change?

Actually, It's not. I'll switch to the new implementation in the following PRs.

File: src/common/meta/src/key/catalog_name.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (WenyXu) [reply src/common/meta/src/key/catalog_name.rs]: Fixed 🥲

**Comment by WenyXu** [comment]
@MichaelScofield PTAL

**Review by MichaelScofield — APPROVED** [review]

**Review by fengjiachun — APPROVED** [review]

File: src/common/meta/src/key.rs [hunk 9]
```diff
@@ -57,18 +65,22 @@ use table_info::{TableInfoKey, TableInfoManager, TableInfoValue};
...
+use self::catalog_name::CatalogNameValue;
+use self::schema_name::SchemaNameValue;
 use crate::error::{InvalidTableMetadataSnafu, Result, SerdeJsonSnafu};
 pub use crate::key::table_route::{TableRouteKey, TABLE_ROUTE_PREFIX};
 use crate::kv_backend::KvBackendRef;
 
 pub const REMOVED_PREFIX: &str = "__removed";
 
-const TABLE_NAME_PATTERN: &str = "[a-zA-Z_:][a-zA-Z0-9_:]*";
+const NAME_PATTERN: &str = "[a-zA-Z_:-][a-zA-Z0-9_:-]*";
```
[the remark this answers is not in the document]
Comment (Fengys123) [reply src/common/meta/src/key.rs:76]: For now we should not allow "-".

**Review by Fengys123 — APPROVED** [review]
Amostly LGTM,  and nit `range_stream` is a bit complicated.

**Pull request merged**